### PR TITLE
feat(visual): add HTML diff report for visual regression failures

### LINKED ISSUE #133: Plan: HTML diff report navegável para falhas de visual regression em PRs
---
plan_id: "002"
title: html-diff-report-for-visual-regression-failures
status: pending
agent: claude
issue: guardiatechnology/design-system#130
parent: guardiatechnology/design-system#130
---

# Plan: HTML diff report navegável para falhas de visual regression em PRs

## Objective

Quando `npm run test-storybook` falha por diff visual no CI, gerar `__image_snapshots__/__diff_report__/index.html` com base × current × diff lado a lado para cada story afetada. Reviewer abre o report sem precisar baixar zip de PNGs soltas.

Refs Tech Task #130 — escopo B. Independente do Plan A (reorganização).

## Steps

- [ ] **1. Worktree + branch**: `chore/{N}-html-diff-report-visual-regression`
- [ ] **2. Avaliação de ferramentas OSS**:
  - [`reg-publish-github-action`](https://github.com/reg-viz/reg-publish-github-action) — pronto, integra com GH Actions
  - Script custom (~80 linhas) — `pixelmatch` (já é dep transitiva de `jest-image-snapshot`) + template HTML simples
  - [`lost-pixel`](https://github.com/lost-pixel/lost-pixel) report mode — provavelmente overkill (puxa Playwright próprio)
- [ ] **3. ADR**: `docs/adr/ADR-{n}-visual-diff-report.md` registrando escolha (qual ferramenta + por quê — peso de deps, complexidade, integração com CI atual)
- [ ] **4. Implementação**: integrar a ferramenta escolhida no `.storybook/test-runner.ts` ou pós-hook do CI
- [ ] **5. Workflow**: `.github/workflows/pull-request.yml` publica `__diff_report__/` como artifact (ou GH Pages preview, conforme decisão da ADR)
- [ ] **6. Bot comment automático** (opcional, depende da ferramenta): comment no PR com link para o report quando o gate visual falha
- [ ] **7. Smoke test**: induzir falha proposital em uma story (cor de um componente), validar que o report aparece com base/current/diff
- [ ] **8. [CONTRIBUTING.md](CONTRIBUTING.md)**: documentar como acessar o report quando o gate visual falha

## Dependencies

- Plan A do #130 **não bloqueia** (reorganização de baselines é ortogonal). Pode ser executado em paralelo ou após.

## Risks

- **Bundle size de OSS tool**: `lost-pixel` puxa Playwright próprio (~150 MB). Mitigação: medir `node_modules` antes/depois; preferir custom script (~80 linhas) se delta for > 30 MB
- **Bot comment requer GH App ou token**: comment automático no PR exige permissão de write. Mitigação: se a ferramenta exige config sensível, ficar só com artifact + link no summary do workflow

## Out of scope

- Self-hosted Argos CI ou similar (over-engineering pro tamanho atual)
- Mudança da estratégia "Git como source of truth" (decidido em #127)
- Aprovação visual via UI tipo Chromatic (decisão registrada em #126 — fica como `commit de baseline` revisado no diff do PR)

### PR DESCRIPTION
**Pull request by fernandoseguim**
## Resumo

Quando o gate visual falha (CI ou local), o reviewer agora abre um `index.html` navegável com `baseline | received | diff` lado a lado por story+tema, em vez de baixar zip e inspecionar PNGs soltas.

Plan #133 fecha o escopo B do Tech Task #130 (escopo A — reorg dos baselines por componente — já mergeou em PR #134).

### Capturar o `received` no test-runner
- `.storybook/test-runner.ts`: adiciona `storeReceivedOnFailure: true` + `customReceivedDir` espelhando a hierarquia de `customDiffDir`. Expõe a renderização atual ao lado do composite diff produzido pelo `jest-image-snapshot`.

### Gerador de report (zero deps)
- `scripts/generate-diff-report.mjs` (~210 linhas, Node stdlib only):
  - Caminha recursivo em `__image_snapshots__/__diff_output__/`
  - Para cada `*-diff.png`, casa com `*-received.png` no mesmo diretório e baseline em `__image_snapshots__/{title}/{theme}/{variant}.png`
  - Gera `__image_snapshots__/__diff_output__/index.html` com cards agrupados por componente
  - Header inclui `branch`, `commit` (7 chars) e link do workflow run via env vars do GH Actions, quando presentes
  - Funciona local também — `node scripts/generate-diff-report.mjs` após uma falha gera o HTML pra inspeção

### Workflow
- `.github/workflows/pull-request.yml`:
  - Novo step `Generate diff report (HTML)` invoca o script após `test-storybook:ci` falhar
  - O HTML vai dentro do mesmo artifact `visual-diffs-${run_id}` (já uploaded)
  - Notice no workflow summary aponta o path

### ADR
- `docs/adr/ADR-001-visual-diff-report-tooling.md`: registra a decisão. Custom script ganhou de `reg-publish-github-action` e `lost-pixel`:
  - `reg-publish` exige duplicar comparador (já temos `jest-image-snapshot`)
  - `lost-pixel` puxa Playwright próprio (~150 MB), conflita com o do test-runner
  - Custom: zero deps novas, alinha com a hierarquia de `__diff_output__/` do Plan #132

### Docs
- `CONTRIBUTING.md`: atualiza fluxo de "atualizar baselines" pra mencionar o HTML, e adiciona seção curta sobre invocação local do gerador

## Acceptance Criteria (Plan #133)

- AC-1: Avaliação de OSS vs custom — ✅ na ADR-001
- AC-2: Implementação no test-runner ou pós-hook CI — ✅ pós-hook no workflow (mais simples; o test-runner produz os PNGs, o script só agrega)
- AC-3: ADR registrando escolha — ✅ ADR-001
- AC-4: Workflow publica report como artifact — ✅ dentro do `visual-diffs-*` existente
- AC-5: Smoke test induzindo falha proposital — ✅ smoke sintético rodado local com fixtures placeholder (2 stories, 2 componentes, validou rendering, paths relativos, e fallback "missing" pra received ausente)
- AC-6: `CONTRIBUTING.md` atualizado — ✅

## Como validar

Smoke test sintético (rodei local antes do commit):

```bash
mkdir -p __image_snapshots__/__diff_output__/components/multi-select/dark
cp __image_snapshots__/components/multi-select/dark/default.png \
   __image_snapshots__/__diff_output__/components/multi-select/dark/default-diff.png
node scripts/generate-diff-report.mjs
# → 2 failing snapshots across 2 components
open __image_snapshots__/__diff_output__/index.html
```

CI vai exercitar o caminho real na próxima falha visual de qualquer PR — gera o HTML automático, junta no artifact.

## Out of scope

- Bot comment automático no PR com link pro report (depende de GH App ou token com write; ficou fora pra evitar config sensível agora — pode ser follow-up se valer)
- Self-hosted UI (Argos, Chromatic) — decisão registrada em #126 e mantida
- Compressão de PNGs no artifact (ortogonal, ganho marginal)

Closes #133
Refs #130

### CODE REVIEW

File: scripts/generate-diff-report.mjs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    const abs = join(dir, entry);
+    const s = statSync(abs);
+    if (s.isDirectory()) {
+      out.push(...walk(abs));
+    } else if (s.isFile()) {
+      out.push(abs);
+    }
+  }
+  return out;
+}
```
[the remark this answers is not in the document]
Comment (fernandoseguim) [reply scripts/generate-diff-report.mjs:69]: Fixed in 5912a70. `walk` agora usa `readdirSync(dir, { withFileTypes: true })`, `statSync` import removido. Uma syscall por entrada em vez de duas — ganho marginal hoje (catalogo pequeno) mas grátis.

File: scripts/generate-diff-report.mjs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    .join("\n");
+  return `<!doctype html>
+<html lang="pt-BR">
+<head>
+<meta charset="utf-8">
+<title>Visual regression diff report (${total})</title>
+<style>
+  :root {
+    --bg: #0e1016;
+    --bg-elev: #1a1d29;
```
[the remark this answers is not in the document]
Comment (fernandoseguim) [reply scripts/generate-diff-report.mjs]: Fixed in 5912a70. `--bg` trocado para Gray 900 (#17171b), reservando Mono Black (#0e1016) para texto/ícone conforme a hierarquia de dark mode do DS. Boa pegada — eu peguei o Mono Black por reflexo, mas Gray 900 separa melhor o fundo do PNG diff (que tem bordas brancas).

File: scripts/generate-diff-report.mjs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+</body>
+</html>
+`;
+}
+
+function renderEmpty() {
+  return `<!doctype html>
+<html lang="pt-BR">
+<head><meta charset="utf-8"><title>Visual regression diff report (0)</title></head>
+<body style="font-family:-apple-system,Roboto,sans-serif;padding:2rem;background:#0e1016;color:#fdfdfd;">
```
[the remark this answers is not in the document]
Comment (fernandoseguim) [reply scripts/generate-diff-report.mjs]: Fixed in 5912a70 — mesmo ajuste aplicado ao `renderEmpty()` inline style.

**Review by fernandoseguim — COMMENTED** [review]
# Argos Review of PR — #135 (commit 935262c)

**Verdict:** 0 BLOCKER, 3 WARNING, 2 NOTE

Re-review on the human reviewer's request via `cry-review-pr`. No prior `CHANGES_REQUESTED` from Argos on this PR → per Publication policy, this is a clean first-touch and posts as `--comment`.

Identity note: `scripts/argos/auth.sh` not present in this repo (`exit 127` observed) — the GitHub App `ahrena-warrior-argos` is not yet wired here, so this review publishes under the human reviewer's identity, not the bot. Bringing the bot online is a separate task (out of scope here).

---

## Axis A — Technical (Node ESM script, Storybook test-runner, GH Actions workflow, ADR)

| Severity | File:Line | Rule | Finding | Suggestion |
|----------|-----------|------|---------|------------|
| WARNING | scripts/generate-diff-report.mjs:86-92 | defensive parsing | `entryFor()` assumes the path under `__diff_output__/` follows `{titleSegments}/{theme}/{variant}-diff.png` and unconditionally treats the second-from-last segment as `theme` and the third-from-last as the deepest title segment. There is no validation that `theme` ∈ `{"light","dark"}`. The `relFromDiff.length < 3` guard only rejects truly shallow paths; a 3-segment path like `foo/bar/x-diff.png` is accepted and silently maps `bar` to `theme`, `foo` to `titleSegments`. This is a latent invariant from `test-runner.ts` (see `THEMES = ["light","dark"]` + `snapshotDirForStory`), not enforced here. If a future change emits a diff outside that hierarchy (e.g., a transient debug PNG, a flat-layout regression), the generator produces a misleading card silently. | Add an explicit `THEMES = new Set(["light","dark"])` and either (a) `return null` when `theme` not in the set (skipped, logged via stderr), or (b) accept it but classify as `theme = "unknown"` in the rendered card. Either path beats silent misclassification. |
| WARNING | scripts/generate-diff-report.mjs:109 | dead branch | `posix.normalize(...split(/[\\/]/).join("/"))` — `relative()` already uses the platform separator and the split+join over `posix.normalize` is redundant. Worse, on POSIX it does nothing useful, and on Windows it converts `\\` to `/` correctly but the trailing `posix.normalize` is a no-op on an already-canonical relative path. The whole `toRel` collapses to `relative(outputDir, abs).split(/[\\/]/).join("/")`. | Simplify to `const toRel = (abs) => relative(outputDir, abs).split(/[\\/]/).join("/");` — keeps cross-platform separator handling, drops the dead `posix.normalize`. |
| NOTE | scripts/generate-diff-report.mjs:146-153 | HTML escaping | `escape()` covers the 5 standard HTML special chars. Story names come from CSF parser of `*.stories.tsx` files in this repo (trusted source) — XSS is not a realistic threat. The escape is correct against the inputs we actually see. No action required; flagging only because the section explicitly asked for a sanity check. | None. |
| NOTE | .github/workflows/pull-request.yml:132-134 | workflow gating | `if: failure() && !regenerate-baselines` fires for any failure in any prior step of the same job (including a11y violation, install error, or build failure). If `__diff_output__/` is empty when the script runs (e.g., a11y fails before any visual comparison), the script falls through `renderEmpty()` and writes a "0 failing snapshots" HTML. That's the safe path and explicitly handled — but the artifact then contains a misleading `index.html` for a non-visual failure. Minor ergonomic issue, not a correctness one. | (Optional) Gate the step on `failure() && steps.test-storybook.conclusion == 'failure'` using a step `id`. Or accept current behavior — `renderEmpty()` is honest enough about the empty state. |

## Axis B — Alignment with Plan #133 ACs and umbrella Issue #130

Plan #133 declares 8 steps + 6 ACs derived. Mapping against the diff:

| AC / Step | Source | Finding | Evidence |
|-----------|--------|---------|----------|
| AC-1 / Step 2 — OSS vs custom evaluation | docs/adr/ADR-001 §"Alternativas avaliadas" | Met. `reg-publish-github-action` (duplicate comparator) and `lost-pixel` (~150 MB Playwright bundle) rejected with concrete trade-offs; custom script accepted with mitigation. | ADR-001 lines 32-49. |
| AC-2 / Step 4 — implementation in test-runner OR post-hook | `.storybook/test-runner.ts` + `.github/workflows/pull-request.yml` | Met. Pre-condition (`storeReceivedOnFailure` + `customReceivedDir`) wired in test-runner; aggregation in CI post-hook. Cleaner split than either-or. | test-runner.ts:192-197 + workflow:132-134. |
| AC-3 / Step 3 — ADR registering the choice | docs/adr/ADR-001 | Met. See Axis F for ADR quality notes. | docs/adr/ADR-001. |
| AC-4 / Step 5 — workflow publishes report as artifact | .github/workflows/pull-request.yml:136-145 | Met. HTML lives inside the existing `visual-diffs-${run_id}` artifact (no second artifact). Notice in summary at line 147-150 documents the path. | workflow:136-145. |
| AC-5 / Step 7 — smoke test inducing failure | PR body §"Como validar" + author note | Met locally via synthetic fixtures (2 entries, 2 components). No CI run has yet exercised the path with a real visual failure on this branch (CI is green). Acceptable — the real-failure path will be exercised on the next PR that breaks a baseline; behavior on `failure()` is deterministic and the empty path is explicitly handled. | PR body. |
| AC-6 / Step 8 — CONTRIBUTING updated | CONTRIBUTING.md +12 lines | Met. New "Diff report HTML local" subsection + updated step 3 in the baseline update flow referencing the HTML. | CONTRIBUTING.md diff. |
| Step 6 — Bot comment on PR with link | Out of scope, declared | Acknowledged. PR body §"Out of scope" lists it with rationale (depends on GH App or write token). Acceptable deferral. | PR body. |

Net: 6/6 ACs met, 1/1 declared out-of-scope item acknowledged. Plan-to-diff alignment is clean.

## Axis C — Local tests

- `npm run typecheck` (tsc -p tsconfig.test.json --noEmit): **clean** (0 errors)
- `npm run lint` (ESLint): **clean** (0 errors, 27 pre-existing warnings unchanged by this PR)
- Edge-case probe against `entryFor()`: see Axis A WARNING #1 — reveals the latent assumption.
- `test-storybook:ci` not re-run locally; the validation CI run `26351720551` already on the head commit reports green per the PR description.

## Axis D — Backward compatibility

- Public package surface: **no change**. `scripts/`, `.github/workflows/`, `.storybook/`, `docs/adr/`, `CONTRIBUTING.md`, and the Calendar story are all dev-tooling / docs. The `dist/` exports from `package.json` are untouched.
- `__image_snapshots__/components/calendar/{light,dark}/default.png` regenerated by the `regenerate-baselines` workflow on this branch (signed commit by GitHub via Git Database API, per the existing flow from Plan #128). Expected and consistent with `lex-visual-regression-ubuntu-sot` (the user's recorded preference + Plan #132 layout).
- `oasdiff` / events schema diff: N/A — no OpenAPI or CloudEvents in this PR.

## Axis E — Security

- Generated HTML is consumed only locally / from the GH Actions artifact (downloaded zip opened in browser). No public surface, no untrusted input — story names come from CSF source files in this repo.
- `escape()` correctly covers the 5 HTML special chars; the only HTML attribute interpolations (`src`, `alt`, `href`) all flow through `escape()`. `loading="lazy"` on ``, `target="_blank" rel="noopener"` on the workflow-run link — both good.
- The `Push regenerated baselines` workflow step keeps `permissions: contents: write` scoped to the `visual-and-a11y` job and guarded by `github.event.pull_request.head.repo.full_name == github.repository` — forks cannot trigger it. Inherited from Plan #128, not changed here.

## Axis F — Lexis / Codex conformance scan

| Severity | File:Line | Lexis / Codex | Finding | Suggestion |
|----------|-----------|---------------|---------|------------|
| WARNING | docs/adr/ADR-001-visual-diff-report-tooling.md | `lex-language-ptbr` + ADR convention | Content is in pt-BR. `language.default` in this project is en (per the framework structure in `.claude/rules/documentation/i18n/`), and the existing baseline ADRs / docs are in English. The body is well-structured (Context, Decision, Alternatives, Consequences, Implementation, Revisita — MADR-aligned with one heading translated) but the language drift means consumers of the framework who read English will need a translation. | (Soft) Either (a) move content to en to match framework default, keeping pt-BR as the localized companion (lex-framework-language pattern), or (b) record in the project's `.ahrena/.directives` that this repo's docs default to pt-BR (if that's a deliberate per-repo choice). Not blocking on its own — the project may have set Portuguese as the local default; verify `language.default` for this repo. |

Other Lexis checks — clean:
- `lex-no-silent-tech-debt`: 1 hit for `revisitado` in `.storybook/test-runner.ts:34` — pre-existing prose ("o hook precisa ser revisitado se..."), not a marker in a code comment. No silent debt added by this PR.
- `lex-frontend-typing`: N/A (the script is `.mjs`, not TypeScript; the test-runner edit is correctly typed).
- `lex-logging-decorator`: 2 `console.log` calls in `generate-diff-report.mjs` (lines 44, 70, 141). This is a build-tool script invoked from a GH Action — not application code. The Lexis exception (3): "global exception handlers at the top of the application" applies by analogy to CLI entrypoints. Acceptable here; would only be a finding if the script grew runtime-app responsibilities.
- `lex-frontend-security`: no `innerHTML` / `dangerouslySetInnerHTML` in the changed surface. The HTML generation uses template strings + `escape()` consistently.
- `lex-design-system-library` / `lex-brand-colors`: the HTML page background uses Gray 900 (`#17171b`) per the gemini follow-up correctly addressed in 5912a70. Inline styles in the report are scoped to a dev-tooling artifact — not a product surface, so the `@guardia/design-system` mandate does not apply.
- `lex-feature-design-docs`: ADR-001 lives at `docs/adr/`, which is outside `docs/{context}/{category}/`. ADRs are explicitly governed by their own convention (`kata-adr-write`), not by the design-doc layout — no violation.

## ADR-001 quality review

The ADR is well-formed against MADR-lite conventions and matches what `kata-adr-write` would produce. Specific notes:

- Front-matter complete (`adr_id`, `title`, `status: accepted`, `date`, `deciders`, `context_issue`, `parent_issue`) — all present.
- Sections present: Context, Decisão, Alternativas avaliadas (3 alternatives with Pró/Contra/Veredito), Consequências (+/− balanced), Implementação, Revisita. Maps cleanly onto MADR.
- "Revisita" section declares 3 concrete triggers (catalog > 500 stories, recurring report gaps, budget for Chromatic/Argos). Strong — most ADRs in the wild drop this.
- Concrete numbers cited (~200 lines, 150 MB, 53 stories × 2 themes = 106). Backs the trade-off claim with data, per `lex-brand-voice`'s "claims backed by evidence" pillar.

Single suggestion: the "Decisão" prose ("script custom em ESM puro... ~200 linhas") would benefit from a one-line summary at the very top of the section for skimmers — currently the reader has to read 4 numbered points to extract the choice. Minor.

## Calendar fix scope question

The Calendar/Default pin (`9a7fd2b`) is scope adjustment of the active Plan — defensible per `lex-agent-focus-on-active-plan` exception (c) "scope adjustment of the active Plan itself". The fix surfaced a latent bug from PR #134 that was actively blocking the visual gate from passing on this branch (Calendar's today-cell highlight ≈ 2k pixels, > 0.2% threshold). Splitting it into a separate Plan would have left this PR's CI red and unmergeable without a baseline override — that's the textbook case for in-scope expansion.

The WHY comment in `Calendar.stories.tsx:26-33` is excellent: explicit lineage (cross-references PR #134), explains the failure mode (react-day-picker semantics of `selected` vs `defaultMonth`), and justifies the pin choice (June 2020 specifically avoids any future "today"). Strong example of `lex-no-silent-tech-debt` exception (i) `# WHY:` documentation.

Defensible.

---

**Next steps:** address 3 WARNINGs in this PR or open follow-up Issues (Axis A WARNING #1 is the most material — latent assumption in `entryFor`; #2 is trivial cleanup; Axis F is a language-default verification). PR is otherwise mergeable.

**Comment by fernandoseguim** [comment]
Argos review endereçado em a1cf94a (rebased to 173c45b):

- **Axis A WARNING #1** — `entryFor()` agora guarda `theme` contra um `THEMES` allowlist (`{light, dark}`). Path fora do shape (ex.: `foo/bar/x-diff.png`) é skipped + logged em stderr. Verifiquei localmente: bad fixture pula com mensagem clara, valid fixture renderiza normal.
- **Axis A WARNING #2** — `toRel` simplificado para `relative(...).split(/[\\/]/).join("/")`. Import `posix` removido.
- **Axis F (language)** — Project `.ahrena/.directives` declara `language.default: pt-BR` e `language.i18n: [pt-BR]` neste repo (override do framework default `en`). ADR-001 em pt-BR está alinhado com o default deste projeto. Sem ação.
- **ADR-001 NOTE** — Adicionado sumário de uma linha no topo da seção «Decisão» para skimmers.
- **HTML escaping NOTE** + **workflow gating NOTE** — sem ação (corretos como estão).
- **Calendar fix scope question** — concordo com o veredito de "defensible". Mantido.

PR vai re-rodar CI; se passar, pronto pra merge.
